File: Leetcode/MaximumAveragePassRatio.rs
// https://leetcode.com/problems/maximum-average-pass-ratio

use std::collections::BinaryHeap;

struct Solution {}

impl Solution {
    pub fn max_average_ratio(mut classes: Vec<Vec<i32>>, extra_students: i32) -> f64 {
        let mut bh = BinaryHeap::<(usize, usize)>::new();

        fn get_delta(class: &Vec<i32>) -> usize {
            let pass = class[0] as usize;
            let total = class[1] as usize;

            let num = 1_000_000_000 * ((pass + 1) * total - pass * (total + 1));
            let den = (total + 1) * total;

            return num / den;
        }

        for (i, class) in classes.iter().enumerate() {
            bh.push((get_delta(class), i));
        }

        for _ in 0..extra_students {
            if let Some((_, i)) = bh.pop() {
                classes[i][0] += 1;
                classes[i][1] += 1;

                bh.push((get_delta(&classes[i]), i));
            }
        }

        let n = classes.len() as f64;

        classes
            .into_iter()
            .map(|class| class[0] as f64 / class[1] as f64)
            .sum::<f64>()
            / n
    }
}

File: Leetcode/MaximumAveragePassRatio.cpp
// https://leetcode.com/problems/maximum-average-pass-ratio

class Solution {
   public:
    double maxAverageRatio(vector<vector<int>>& classes, int extraStudents) {
        const auto getDelta = [&](const auto& cl, const auto& ind) {
            const long long pass = cl[0];
            const long long total = cl[1];
            const auto num = ((pass + 1) * total - pass * (total + 1));
            const auto den = (total + 1) * total;

            return pair<double, size_t>{(1. * num / den), ind};
        };

        auto ind = 0;
        vector<pair<double, size_t>> data;
        data.reserve(classes.size());

        transform(classes.cbegin(), classes.cend(), std::back_inserter(data),
                  [&](const auto& cl) { return getDelta(cl, ind++); });

        priority_queue<pair<double, size_t>, vector<pair<double, size_t>>,
                       std::less<>>
            minHeap(less<>(), move(data));

        while (extraStudents-- > 0) {
            const auto [_, ind] = minHeap.top();
            minHeap.pop();

            ++classes[ind][0];
            ++classes[ind][1];

            minHeap.push(getDelta(classes[ind], ind));
        }

        double avg = 0;
        while (!minHeap.empty()) {
            const auto [_, ind] = minHeap.top();
            minHeap.pop();
            avg += 1. * classes[ind][0] / classes[ind][1];
        }

        return avg / classes.size();
    }
};